sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryMetrics.java

```diff
@@ -0,0 +1,907 @@
+import com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdEndpoint;
+import com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdEndpointStatistics;
+import com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdMetricsCompletionRecorder;
+import com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdRequestRecord;
+import com.azure.cosmos.implementation.guava25.net.PercentEscaper;
+import com.azure.cosmos.implementation.query.QueryInfo;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class ClientTelemetryMetrics {
+    private static final Logger logger = LoggerFactory.getLogger(ClientTelemetryMetrics.class);
+    private static final ImplementationBridgeHelpers.CosmosAsyncClientHelper.CosmosAsyncClientAccessor clientAccessor =
+        ImplementationBridgeHelpers.CosmosAsyncClientHelper.getCosmosAsyncClientAccessor();
+    private static final
+        ImplementationBridgeHelpers.CosmosDiagnosticsHelper.CosmosDiagnosticsAccessor diagnosticsAccessor =
+            ImplementationBridgeHelpers.CosmosDiagnosticsHelper.getCosmosDiagnosticsAccessor();
+    private static final PercentEscaper PERCENT_ESCAPER = new PercentEscaper("_-/.", false);
+
+    private static CompositeMeterRegistry compositeRegistry = createFreshRegistry();
+    private static final ConcurrentHashMap<MeterRegistry, AtomicLong> registryRefCount = new ConcurrentHashMap<>();
+
+    private static String convertStackTraceToString(Throwable throwable)
+    {
+        try (StringWriter sw = new StringWriter();
+             PrintWriter pw = new PrintWriter(sw))
+        {
+            throwable.printStackTrace(pw);
+            return sw.toString();
+        }
+        catch (IOException ioe)
+        {
+            throw new IllegalStateException(ioe);
+        }
+    }
+
+    private static CompositeMeterRegistry createFreshRegistry() {
+        CompositeMeterRegistry registry = new CompositeMeterRegistry();
+        if (logger.isTraceEnabled()) {
+            registry.config().onMeterAdded(
+                (meter) -> {
+                    logger.trace(
+                        "Meter '{}' added. Callstack: {}",
+                        meter.getId().getName(),
+                        convertStackTraceToString(new IllegalStateException("Dummy")));
```

Comment:
why not using Thread.getCurrentStackTrace()

Reply:
That method doesn't exist. You could use Thread.currentThread().getStackTrace() - but then would need to walk down the StackTraceELements and put them in a string - leaving you do decide which line feeds to use etc. - this approach simplifies this by making sure the formatting is identical to the one in exceptions - and since this is deifnitely not on the hot path efficiency is not a concern)